scripts/generate-shaders.js

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env node
+
+var path = require('path');
+var fs = require('fs');
+
+var inputPath = 'mapbox-gl-js/shaders';
+var outputPath = 'src/mbgl/shaders';
+
+[
+'circle',
+'collision_box',
+'debug',
+'fill',
+'fill_outline',
+'fill_outline_pattern',
+'fill_pattern',
+'line',
+'line_pattern',
+'line_sdf',
+'raster',
+'symbol_icon',
+'symbol_sdf'
+].forEach(function (shaderName) {
+    function applyPragmas(source, pragmas) {
+        return source.replace(/#pragma mapbox: ([\w]+) ([\w]+) ([\w]+) ([\w]+)/g, (match, operation, precision, type, name) => {
+            return pragmas[operation]
+                .join("\n")
+                .replace(/\{type\}/g, type)
+                .replace(/\{precision\}/g, precision)
+                .replace(/\{name\}/g, name);
+        });
+    }
+
+    function vertexSource() {
+        var prelude = fs.readFileSync(path.join(inputPath, '_prelude.vertex.glsl'));
+        var source = fs.readFileSync(path.join(inputPath, shaderName + '.vertex.glsl'), 'utf8');
+        return prelude + applyPragmas(source, {
+                define: [
+                    "uniform lowp float a_{name}_t;",
+                    "attribute {precision} {type} a_{name}_min;",
+                    "attribute {precision} {type} a_{name}_max;",
+                    "varying {precision} {type} {name};"
+                ],
+                initialize: [
+                    "{name} = mix(a_{name}_min, a_{name}_max, a_{name}_t);"
+                ]
+            });
+    }
+
+    function fragmentSource() {
+        var prelude = fs.readFileSync(path.join(inputPath, '_prelude.fragment.glsl'));
+        var source = fs.readFileSync(path.join(inputPath, shaderName + '.fragment.glsl'), 'utf8');
+        return prelude + applyPragmas(source, {
+                define: [
+                    "varying {precision} {type} {name};"
+                ],
+                initialize: [
+                ]
+            });
+    }
+
+    var content = "#pragma once\n" +
+    "\n" +
+    "// NOTE: DO NOT CHANGE THIS FILE. IT IS AUTOMATICALLY GENERATED.\n" +
+    "\n" +
+    "#include <mbgl/gl/gl.hpp>\n" +
+    "\n" +
+    "namespace mbgl {\n" +
+    "namespace shaders {\n" +
+    "\n" +
+    "class " + shaderName + " {\n" +
+    "public:\n" +
+    "    static constexpr const char* name = \"" + shaderName + "\";\n" +
+    "    static constexpr const char* vertexSource = R\"MBGL_SHADER(\n" + vertexSource() + ")MBGL_SHADER\";\n" +
+    "    static constexpr const char* fragmentSource = R\"MBGL_SHADER(\n" + fragmentSource() + ")MBGL_SHADER\";\n" +
+    "};\n" +
+    "\n" +
+    "} // namespace shaders\n" +
+    "} // namespace mbgl\n";
+
+    fs.writeFileSync(path.join(outputPath, shaderName + '.hpp'), content);
```

Comment:
Let's use a pattern that conditionally writes the shader code if changes occurred, like in [style-code.js](https://github.com/mapbox/mapbox-gl-native/blob/master/scripts/style-code.js#L25). Previously, we were running this script for every shader file, while we are now running it once for every shader file. Previously, the build system tracked changes to shader files and only rebuild the files when that particular file changed. Rewriting all files (even if they have the same file content) will trigger a rebuild that takes time. Changes to shaders are necessary when debugging rendering issues or implementing new features or adding new shaders.